src/test/regress/expected/alter_table.out

```diff
@@ -2781,19 +2783,19 @@ select * from my_locks order by 1;
 commit;
 begin; alter table alterlock set (fillfactor = 100);
 select * from my_locks order by 1;
-  relname  |       max_lockmode       
------------+--------------------------
- alterlock | ShareUpdateExclusiveLock
- pg_toast  | ShareUpdateExclusiveLock
+  relname  |    max_lockmode     
```

Comment:
I cannot match the test case with c code change. Is it just a test change? Can you explain a bit why the test change?

Reply:
Yes, this is a different behaviour between PG and GP. It's described in the main PR body:

- Different behaviours in `AlterTableGetLockLevel / AlterTableGetRelOptionsLockLevel`:

In case `AT_ResetRelOptions`, the lock mode is determined in `AlterTableGetRelOptionsLockLevel()`, Setting multiple options together uses the highest lock level required for any option. And in GPDB, the highest lockmode is set by this relOption:

```
(gdb) p *relOpts[i]
$7 = {name = 0xe946ef "autosummarize", desc = 0xe9b82a "Enables automatic summarization on this BRIN index", kinds = 1024, lockmode = 8, namelen = 13, type = RELOPT_TYPE_BOOL}
```

which is AccessExclusiveLock.

Reply:
@Aegeaner from the commit message, it just states why the difference.

To the problem itself, what lock mode should it hold and why?

Reply:
It should be `AccessExclusiveLock` as it currently is, as there is a pre-defined relOption for `autosummarize`:
https://github.com/greenplum-db/gpdb-postgres-merge/blob/iteration_REL_12_12/src/backend/access/common/reloptions.c#L106~L111

```
{
			"autosummarize",
			"Enables automatic summarization on this BRIN index",
			RELOPT_KIND_BRIN,
			AccessExclusiveLock
		},
```

The details are in commit: https://github.com/greenplum-db/gpdb-postgres-merge/commit/7526e10224f0792201e99631567bbe44492bbde4

Reply:
From the comments in `src/backend/access/common/reloptions.c`:

https://github.com/greenplum-db/gpdb-postgres-merge/blob/813dec566513d76a16d9e27d0bf06cfeb1d4c0e7/src/backend/access/common/reloptions.c#L62

> The default choice for any new option should be AccessExclusiveLock.
> In some cases the lock level can be reduced from there, but the lock
> level chosen should always conflict with itself to ensure that multiple
> changes aren't lost when we attempt concurrent changes.
> The choice of lock level depends completely upon how that parameter
> is used within the server, not upon how and when you'd like to change it.
> Safety first. Existing choices are documented here, and elsewhere in
> backend code where the parameters are used.
> 
> In general, anything that affects the results obtained from a SELECT must be
> protected by AccessExclusiveLock.

This is also aligned with upstream.

Reply:
I test this under Postgres, Postgres use `ShareUpdateExclusiveLock`. I still do not understand why Greenplum has to use `AccessExclusiveLock`.

Reply:
@kainwen Sorry the previous explanation was wrong, the lockmode was updated by the `fillfactor` option:
```
(gdb) p *relOpts[i]
$4 = {name = 0x23bed70 "fillfactor", desc = 0x23bed98 "Packs bitmap index pages only to this percentage", kinds = 4096, lockmode = 8, namelen = 10,
  type = RELOPT_TYPE_INT}
```
This is a GP-only reloption and the default value is `ShareUpdateExclusiveLock`, there should have been a lock upgrading happened somewhere:
https://github.com/greenplum-db/gpdb-postgres-merge/blob/d789f7d64aea5863d04be0805e7d9cfb94941a53/src/backend/access/common/reloptions_gp.c#L68